constellations/orion/src/controllers/classes_by_semesters_controller.go

```diff
@@ -0,0 +1,125 @@
+package controllers
+
+import (
+	"net/http"
+
+	"github.com/ahsu1230/mathnavigatorSite/constellations/orion/src/domains"
+	"github.com/ahsu1230/mathnavigatorSite/constellations/orion/src/repos"
+	"github.com/gin-gonic/gin"
+)
+
+func GetAllProgramsSemestersClasses(c *gin.Context) {
+	var classSemester domains.ProgramClassesBySemester
+	var listResults []domains.ProgramClassesBySemester
+
+	// Fetch progrmas, semesters, classes from repo functions
+	publishedOnly := ParseParamPublishedOnly(c)
+
+	programs, err := repos.ProgramRepo.SelectAll(publishedOnly)
+	semesters, err := repos.SemesterRepo.SelectAll(publishedOnly)
+	classes, err := repos.ClassRepo.SelectAll(publishedOnly)
+
+	// Convert lists into maps
+	programMap := make(map[string]domains.Program)
+	semesterMap := make(map[string]domains.Semester)
+	classMap := make(map[string]domains.Class)
+
+	for i := 0; i < len(programs); i++ {
+		programMap[programs[i].ProgramId] = programs[i]
+	}
+	for i := 0; i < len(semesters); i++ {
+		semesterMap[semesters[i].SemesterId] = semesters[i]
+	}
+	for i := 0; i < len(classes); i++ {
+		classMap[classes[i].ClassId] = classes[i]
+	}
+
+	// Semester to class list map (maps semesterId to Class object)
+	semesterClassMap := make(map[string][]domains.Class)
+
+	// Loop over semesterIds
+	for i := 0; i < len(semesters); i++ {
```

Comment:
Here's a suggestion to make this for loop overall easier.

```
var listResults []domains.ProgramClassesBySemester
for _______________ {
    semesterObj := semesters[i]
    programClasses := createProgramClassesForSemester(semestersObj.SemesterId)
    programClassesBySemester := domains.ProgramClassesBySemester{
		Semester:       semester,
		ProgramClasses: programClasses,
	}
    listResults = append(listResults, programClassesBySemester)
}
```

Now it's super easy to see what's going on in this layer.
You are mapping over a list, grabbing the semester. Somewhere you create a bunch of programClasses and bundle it together with this class.

Reply:
From there, let's define the createProgramClassesForSemester(...) - there's some logic differences that I think would make your code work faster and neater.

```
func createProgramClassesForSemester(semesterId, ... // other stuff like the classMaps or programMaps probably) {
    // define list of []ProgramClass
    // find classes with semesterId

    // Step A. This can be its own function too!
    // create a map of programId -> list of classes
    // for each class found, 
             extract programId
             add to map (programId -> list of classes (with new class))

    // Step B. Using the map from step A This might be its own function too!
    define an empty list of []ProgramClass
    // for each (program) in (list of programs from original programRepo.SelectAll)
         // 1. get list of classes for this programId
         // 2. for each class in this list, 
                       extract class from ClassMap (classObj)
                       create a ProgramClass from programObject and classObject
                       append to a list of []ProgramClass
   return []ProgramClass
}
```

Something like that ^. I might've missed some details though. But I think changing your functionality to use the part I labeled in step A will help a lot.

The overall idea here is to not have too many indents in a function. We call this modularization. A lot of people say 80 character limit to each line of code in order to keep it readable. If half of that is indents, you're wasting a lot of space and it's hard to code, so you should refactor things out into a separate function - and have more of that complicated logic isolated into functions.